teps/0114-custom-tasks-beta.md

```diff
@@ -0,0 +1,264 @@
+beta.
+
+## Motivation
+
+[TEP-0002: Custom Tasks][tep-0002] introduced `Custom Tasks` which allow users to extend the functionality of Tekton 
+Pipelines. Users implement controllers that watch for `Runs` that implement their types of `Custom Tasks`. 
+
+`Runs` is in `v1alpha1`. `Runs` and `PipelineResources` are the only remaining types in `v1alpha1` - see [TEP-0105]
+[tep-0105] for more details on `v1alpha1`.
+
+Use of `Custom Tasks` in `Pipelines` is also gated behind feature flags: either `enable-custom-tasks` needs to be 
+set to `"true"` or `enable-api-fields` needs to be set to `"alpha"`.
+
+`Custom Tasks` have been available since [v0.19.0][v0.19.0] of Tekton Pipelines released in December 2020.
+Many `Custom Tasks` have been created to extend Tekton Pipelines functionality, including but not limited to:
+- [Task Loops][task-loops]: Runs a `Task` in a loop with varying `Parameter` values. 
+- [Pipeline Loops][pipeline-loops]: Runs a `Pipeline` in a loop with varying `Parameter` values.
+- [Common Expression Language][cel]: Provides Common Expression Language support in Tekton Pipelines.
+- [Wait][wait]: Waits a given amount of time, specified by a `Parameter` named "duration", before succeeding.
+- [Approvals][approvals]: Pauses the execution of `PipelineRuns` and waits for manual approvals.
+- [Pipelines in Pipelines][pipelines-in-pipelines]: Defines and executes a `Pipeline` in a `Pipeline`. 
+- [Task Group][task-group]: Groups `Tasks` together as a `Task`.
+- [Pipeline in a Pod][pipeline-in-pod]: Runs `Pipeline` in a `Pod`.
+
+As discussed in [TEP-0096][tep-0096], `Custom Tasks` are critical for extensibility in Tekton Pipelines, as such
+promoting `Custom Tasks` to beta is a blocker for releasing v1 of Tekton Pipelines.
+
+## Proposal
+
+In this section, we scope the promotion of `Custom Tasks` to beta. The work in [required](#required) must be done,
+while the work in [optional](#optional) may be done.
+
+### Required
+
+The work scoped here are blockers for promotion.
+
+#### API Changes
+
+##### v1alpha1 to v1beta1 + Run to CustomRun
+
+Move `v1alpha1.Run` to `v1beta1.CustomRun`.
+
+The migration from `v1alpha1` to `v1beta1` provides an opportunity to rename `Runs` to `CustomRuns`, as discussed in
+[tektoncd/pipelines#5005][5005] and [API WG meeting][wg-notes].
+
+##### Embedded Specifications
+
+Simplify embedded specifications in `Runs`.
+
+Today, [embedded specs][runs-spec-docs] are in `Run.Spec.Spec.Spec`, for example:
+
+  ```yaml
+  apiVersion: tekton.dev/v1alpha1
+  kind: Run
+  metadata:
+    name: embedded-run
+  spec:
+    spec:
+      apiVersion: example.dev/v1alpha1
+      kind: Example
+      spec:
+        field1: value1
+        field2: value2
+  ```
+
+Simplify by removing the extra `spec` layer such that it's `Run.Spec.Spec`, for example:
+
+  ```yaml
+  apiVersion: tekton.dev/v1alpha1
+  kind: Run
+  metadata:
+    name: embedded-run
+  spec:
+    apiVersion: example.dev/v1alpha1
+    kind: Example
```

Comment:
forgot to add the double spec as indicated in line 105 - will fix shortly

re: the common interface, that also came up when discussing pipelines in pipelines - agreed that we should explore doing that, but would prefer to keep it out of scope for custom tasks beta (will open an issue to track this as it has come up several times)

Reply:
> one problem here is that `Run.Spec.Ref` also includes API version and Kind. would it make sense to instead do:
> 
> ```yaml
> apiVersion: tekton.dev/v1alpha1
> kind: Run
> metadata:
>   name: embedded-run
> spec:
>   spec:
>     apiVersion: example.dev/v1alpha1
>     kind: Example
>     field1: value1
>     field2: value2
> ```
> 
> ? It's a bit weird though since it lumps metadata in with Spec.

I'd rather we don't lump metadata with spec, and this will be inconsistent with the rest of our API

I now think that we should leave `run.spec.spec.spec` for custom tasks -- it doesn't seem terrible compared to lumping the metadata -- just a bit lengthy but it's what we need

> Or should we try to create a common interface for a "ref" (to be used for PipelineRef, TaskRef, and CustomTaskRef)? We have "APIVersion" in both PipelineRef and TaskRef right now but don't really use it, and we have "Kind" for TaskRef but (I think) it must be "Task" for tasks.

this also came up when discussing pipelines in pipelines - it's not clear how this would solve our problem here, could you provide an example please?

but I agree that we should explore doing that, but would prefer to keep it out of scope for custom tasks beta (will open an issue to track this as it has come up several times)

Reply:
> I'd rather we don't lump metadata with spec, and this will be inconsistent with the rest of our API

Yeah on second thought I agree

> could you provide an example please?

Here's a TaskRef today:
```
type TaskRef struct {
	Name string `json:"name,omitempty"`
	Kind TaskKind `json:"kind,omitempty"`
	APIVersion string `json:"apiVersion,omitempty"`
	Bundle string `json:"bundle,omitempty"`
	ResolverRef `json:",omitempty"`
}
```
We could instead have:
```
type Ref struct {
	Name string `json:"name,omitempty"`
	ResolverRef `json:",omitempty"`
}
```
Here, a ref is assumed to be a TaskRef in a TaskRun and a PipelineRef in a PipelineRun (we don't use apiversion that I'm aware of). A CustomRun would always have an APIVersion and Kind under "CustomRun.Spec" (instead of putting it into "CustomRun.Spec.Spec" or "CustomRun.Spec.Ref").

I think this is similar to your suggestion of this:

```
# before
apiVersion: tekton.dev/v1alpha1
kind: Run
metadata:
  name: embedded-run
spec:
  ref:
    apiVersion: example.dev/v1alpha1
    kind: Example
    name: my-example
---
# after
apiVersion: tekton.dev/v1alpha1
kind: Run
metadata:
  name: embedded-run
spec:
  apiVersion: example.dev/v1alpha1
  kind: Example
  name: my-example
```

However I just want to point out that if we decide to go with this solution we'd probably want to remove APIVersion and Kind from TaskRef.

Reply:
since my comment, I updated the suggestion to another approach - see the TEP for details

Reply:
Tahnks! I will bring this to the issue you opened, thanks for creating that